torch/nn/utils/_expanded_weights/linear_expanded_weights.py

```diff
@@ -0,0 +1,35 @@
+import torch
+import torch.nn.functional as F
+from .expanded_weights_impl import implements_per_sample_grads
+from .expanded_weights_utils import \
+    forward_helper, set_grad_sample_if_exists, grad_if_exists_for_input, unpack_expanded_weight_or_tensor
+
+@implements_per_sample_grads(F.linear)
+class LinearPerSampleGrad(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, *expanded_args):
+        if len(expanded_args[0].shape) <= 1:
+            raise RuntimeError("Input does not have a batch dimension. Expanded Weights "
+                               f"expected input of at least rank 2, got of rank {len(expanded_args[0].shape)}")
+        expanded_args = expanded_args[:-1]
+        expanded_kwargs = {'bias': expanded_args[2] if len(expanded_args) == 3 else None}
+        expanded_args_without_kwargs = expanded_args[:2]
+        output, aux_outputs = forward_helper(F.linear, expanded_args_without_kwargs, expanded_kwargs, 1)
+        ctx.args = expanded_args_without_kwargs
+        ctx.kwargs = expanded_kwargs
+        ctx.aux_outputs = aux_outputs
```

Comment:
Okay the reason why I think these lines tripped me up is that they are generic and that makes them difficult to read.

Roughly speaking, here's what we're doing in the forward pass of this (and all autograd.Function for per-sample-grads):
- run f(*unexpanded_args, *unexpanded_kwargs) (or some hacked version of f, if we need intermediate values)
- save values for backward: save the required unexpanded arguments and intermediates
- return the output (as opposed to all intermediate values)

Now, the reason why this code is confusing is that it's not clear what is being saved. We are saving all the args and kwargs, but we're also saving "aux outputs", which turn out to be nothing in this case.

It would make sense for this to be generic if we planned to refactor all of the autograd.Function forward passes to look the same. Is that a good idea?

For F.linear -- there is actually no need to save bias for the backward pass, and if input does not require gradient, then there is no need to save the input! (doesn't need to happen in this PR, but those are potential optimizations). So we might not want all the autograd.Function forward passes to look similar (especially because there are already checks specific to F.linear here)

If we decide that we want this to look generic, I'd probably recommend aux_outputs be renamed to intermediates.

If we decide we don't want this to look generic, It could read better as:
```
output, = forward_helper(F.linear, expanded_args, expanded_kwargs)
ctx.unexpanded_weight = expanded_args[1]
return output
```
For e.g. group_norm this could look like:
```
output, mean, rstd = forward_helper(torch.aten.ops.native_group_norm, expanded_args, expanded_kwargs)
ctx.mean = mean
ctx.rstd = rstd
return output
```
The benefit of the non-generic form is that one doesn't have to go deep diving into the backward() part of the autograd.Function to see exactly what args, kwargs, aux_outputs are.